ts/WoltLabSuite/Core/Ui/Message/Share/Dialog.ts

```diff
@@ -0,0 +1,170 @@
+/**
+ * Shows the share dialog when clicking on the share button of a message.
+ *
+ * @author  Matthias Schmidt
+ * @copyright  2001-2021 WoltLab GmbH
+ * @license  GNU Lesser General Public License <http://opensource.org/licenses/lgpl-license.php>
+ * @module  WoltLabSuite/Core/Ui/Message/Share/Dialog
+ */
+
+import UiDialog from "../../Dialog";
+import DomUtil from "../../../Dom/Util";
+import * as DomTraverse from "../../../Dom/Traverse";
+import * as Language from "../../../Language";
+import * as Clipboard from "../../../Clipboard";
+import * as UiNotification from "../../Notification";
+import * as StringUtil from "../../../StringUtil";
+import DomChangeListener from "../../../Dom/Change/Listener";
+import * as UiMessageShare from "../Share";
+import * as UiMessageShareProviders from "./Providers";
+
+const shareButtons = new WeakSet<HTMLElement>();
+
+/**
+ * Copies the contents of one of the share dialog's input elements to the clipboard.
+ */
+async function copy(event: Event): Promise<void> {
+  event.preventDefault();
+
+  const target = event.currentTarget as HTMLElement;
+  const input = DomTraverse.prevBySel(target, 'input[type="text"]') as HTMLInputElement;
+
+  await Clipboard.copyTextToClipboard(input.value);
+
+  UiNotification.show(Language.get("wcf.message.share.copy.success"));
+}
+
+/**
+ * Returns all of the dialog elements shown in the dialog.
+ */
+function getDialogElements(shareButton: HTMLElement): string {
+  let dialogOptions = "";
+
+  let permalink = "";
+  if (shareButton instanceof HTMLAnchorElement && shareButton.href) {
+    permalink = shareButton.href;
+  }
+
+  if (permalink) {
+    dialogOptions += getDialogElement("wcf.message.share.permalink", permalink);
+  }
+  if (shareButton.dataset.bbcode) {
+    dialogOptions += getDialogElement("wcf.message.share.permalink.bbcode", shareButton.dataset.bbcode);
+  }
+  if (permalink && shareButton.dataset.linkTitle) {
+    if (!shareButton.dataset.bbcode) {
+      dialogOptions += getDialogElement(
+        "wcf.message.share.permalink.bbcode",
+        `[url='${StringUtil.escapeHTML(permalink)}']${StringUtil.escapeHTML(shareButton.dataset.linkTitle)}[/url]`,
```

Comment:
HTML escaping for BBCodes probably is not that useful / incorrect. It'll probably break any links containing an `&`.